### DISCUSSION TITLE
Querying arbitrary DBs using DuckDB beyond Postgres & SQLite

### DISCUSSION THREAD

**Discussion by voberoi** [duckdb/duckdb#6645] [Ideas]
Tagging @Alex-Monahan and @hannes: https://twitter.com/__AlexMonahan__/status/1633561419580063744.

Also cc @jwills: https://twitter.com/josh_wills/status/1633640607641133056

@Alex-Monahan brought up the idea in this tweet: https://twitter.com/__AlexMonahan__/status/1633532829341736961.

We have the Postgres & SQLite scanner extensions, but I would be delighted to be able to query and pull data from arbitrary databases using DuckDB. Being able to do this via ODBC, as Alex suggested, seems like a fast path to doing so.

I get gobs of benefit from sticking to DuckDB's query dialect and environment regardless of where I'm pulling data. Today, I use CSV or JSON exports from external systems to do this.

But I'd like to skip that step and go straight to the source -- creating views or materializing tables using fresh data from external DBs. I get more control this way and it's one more step I can eliminate from the process (exporting files and dumping them somewhere).

I'm keen to help with this to the extent I can.

**Comment by jwills** [comment]
Chiming in here, also very happy to help with this: my desire is to be able to use `dbt-duckdb` as a complete solution for ETLing data from upstream source systems into a DuckDB database.

My initial use case for this is a simple approach to implementing [data contracts](https://benn.substack.com/p/data-contracts) as an integration test between an upstream database and downstream data pipelines that we can run via CICD _before_ we push upstream changes to production to ensure that we don't break the downstream data pipelines; this is already possible today for some teams thanks to the Postgres scanner and I would be keen to extend this power to data teams using MySQL, SQLServer, etc., etc.

**Reply by voberoi** [reply]
Yes! These are all things I want too!

I am a `dbt-duckdb` megafan, Josh.

**Reply by sebastian9610** [reply]
I like the idea, I currently work with different data sources such as Redshift, Oracle, SQL Server, csv and having everything unified in DuckDB sounds great, thus saving me the time to go through pandas that triggers my ram memory. I have looked to see that adbc and odbc supports were included but I can't find a clear way to replicate the connection so I think they are not ready yet.

**Comment by dacort** [comment]
As somebody that just built a rust-based extension to load data from Amazon Athena I’d love something to make this easier. 😂 I did not enjoy having to reimplement “standard” things like fetching column metadata and even results using the API. 

That said there’s still a fair bit of glue that needs to happen in scenarios like this. I also built the Athena driver for Metabase - it supports JDBC drivers, but there were lots of tweaks that still had to me made including:
- Coercing data types
- Timezones, time zones, and offsets oh my
- Functionality supported in some DBs vs others


That said, being able to rely on the official JDBC driver definitely made things a *lot* easier.

**Comment by alexey-milovidov** [comment]
ClickHouse supports this through ODBC and JDBC bridges: https://clickhouse.com/docs/en/engines/table-engines/#integration-engines

**Comment by mskyttner** [comment]
Attaching to data provided over an ADBC bridge maybe would be powerful for accessing columnar data sources? https://arrow.apache.org/blog/2023/01/05/introducing-arrow-adbc/

**Comment by rupurt** [comment]
I've started a Db2 scanner via ODBC based on the Postgres scanner https://github.com/fremantle-industries/db2_scanner. Would love to collaborate with folks as I'm not a C++ programmer but looking forward to have a reason to learn it :smile:

**Comment by abubelinha** [comment]
I can't be of help, but I'd love to see this implemented too ([my use case](https://github.com/duckdb/duckdb/discussions/13141))

I think [this discussion](https://github.com/duckdb/duckdb/discussions/2728#discussioncomment-6716184) should be linked, since they talk about this:
https://github.com/rupurt/odbc-scanner-duckdb-extension

Thanks in advance @rupurt !!
@abubelinha

**Comment by Hunknown** [comment]
I wonder if it couldn't be implemented through [GDAL](https://gdal.org/en/latest/drivers/vector/odbc.html), as it is already used for [Excel file import/export](https://duckdb.org/docs/guides/file_formats/excel_import) in the [spatial extension](https://duckdb.org/docs/extensions/spatial/overview.html). Note that GDAL only supports read-only query for now though.

`FROM ST_Drivers()` shows that duckdb currently doesn't support this specific driver, so I guess some work would be needed for it to be compatible with duckdb.

**Comment by rupurt** [comment]
When the new C bindings with catalog support are released I'll pick the odbc scanner extension back up. It was becoming too much effort figuring out how to keep up with it and how to make C binding wrappers for Zig.
